Kernel/Devices/Loop/LoopDevice.cpp

```diff
@@ -0,0 +1,186 @@
+        if (deref_base())
+            return false;
+        m_list_node.remove();
+        const_cast<LoopDevice&>(*this).revoke_weak_ptrs();
+        return true;
+    });
+    if (did_hit_zero) {
+        const_cast<LoopDevice&>(*this).will_be_destroyed();
+        delete this;
+    }
+    return did_hit_zero;
+}
+
+static ErrorOr<NonnullRefPtr<LoopDevice>> create_or_find_with_already_attached_inode(LoopDevice::ExposedList& locked_all_exposed_instances, LoopDevice::List& locked_all_instances_list, Custody& custody)
+{
+    // NOTE: Don't allow creating a new device if we already have such
+    // device with associated Inode, as it will make it possible to create
+    // multiple devices with the same Inode, which is bad for us.
+    // The main problem with this is that different FileSystem instances
+    // can be created that will be directed to the same Inode, therefore
+    // creating a possible filesystem corruption.
+    auto& inode = custody.inode();
+    for (auto& device : locked_all_instances_list) {
+        if (device.inode().identifier() == inode.identifier())
+            return device;
+    }
+    auto device = TRY(DeviceManagement::try_create_device<LoopDevice>(custody, s_loop_device_id.fetch_add(1)));
+    locked_all_instances_list.append(*device);
+    locked_all_exposed_instances.append(*device);
+    return *device;
+}
+
+ErrorOr<NonnullRefPtr<LoopDevice>> LoopDevice::create_with_file_description(OpenFileDescription& description)
+{
+    auto custody = description.custody();
+    if (!custody)
+        return Error::from_errno(EINVAL);
+
+    // NOTE: We only support regular inode files, because anything else
+    // just doesn't make sense (could be non-seekable files or char devices)
+    if (!custody->inode().metadata().is_regular_file())
+        return Error::from_errno(ENOTSUP);
+
+    // NOTE: We use the inode watcher support as indication on whether we should
+    // allow the user to create a loop device from a given file.
+    // We could technically allow the user to create a loop device from a file
+    // on SysFS, ProcFS, etc, but the added value from allowin this is non-existent
+    // because there's simply no good reason to ever do this kind of operation.
+    //
+    // Also, adding a watcher on an Inode that is now backing a loop device
+    // is a **good way to manage and track filesystem changes** on that file,
+    // so we would want to ensure that userspace can always expect to do this
+    // on the Inode that is attached to a loop device.
+    //
+    // If you need more justification, then filesystems that don't support watchers
+    // for the most part are not keeping Inode objects and instead keep re-creating
+    // them - this has serious consequences on the integrity of loop devices, as we
+    // rely on the backing Inode to be immutable while the LoopDevice is alive.
+    if (!custody->inode().fs().supports_watchers() || !custody->inode().fs().supports_backing_loop_devices())
```

Comment:
I don't know about this whole inode watcher support story: since we manage the entire implementation of all filesystems, why not just implement `.supports_backing_loop_devices()` correctly and leave out `.supports_watchers()`? There's no technical reason watchers should be supported and if a filesystem has some drawbacks that prohibit it from supporting loop devices (as described in your comment), that logic should go into `.supports_backing_loop_devices()` intead.